### Prediction of Employee Attrition/Burnout
Authors: Florencia Luque and Simon Schmetz

In [32]:
import pandas as pd
import numpy as np
import sklearn as skn


In [33]:
data_available = pd.read_csv("attrition_availabledata_10.csv.gz", compression='gzip')
data_competetion = pd.read_csv("attrition_competition_10.csv.gz", compression='gzip')


In [34]:
data_available.head()

hrs  absences  JobInvolvement  PerformanceRating  \
0  7.462297      13.0             3.0                3.0   
1  7.488597       6.0             1.0                3.0   
2  7.116541      16.0             2.0                4.0   
3  9.709125       6.0             3.0                4.0   
4  6.297729       6.0             4.0                3.0   

   EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance   Age  \
0                      3.0              3.0              3.0  48.0   
1                      3.0              4.0              4.0  31.0   
2                      1.0              1.0              3.0  36.0   
3                      3.0              3.0              3.0  49.0   
4                      3.0              2.0              3.0  40.0   

      BusinessTravel              Department  ...  Over18  PercentSalaryHike  \
0      Travel_Rarely  Research & Development  ...       Y               13.0   
1  Travel_Frequently  Research & Development  ...       Y               17.0   
2      Travel_Rarely  Research & Development  ...       Y               24.0   
3      Travel_Rarely  Research & Development  ...       Y               22.0   
4  Travel_Frequently  Research & Development  ...       Y               13.0   

  StandardHours  StockOptionLevel  TotalWorkingYears TrainingTimesLastYear  \
0           8.0               1.0               18.0                   2.0   
1           8.0               1.0                3.0                   2.0   
2           8.0               0.0                6.0                   6.0   
3           8.0               0.0               25.0                   2.0   
4           8.0               1.0                8.0                   3.0   

   YearsAtCompany YearsSinceLastPromotion YearsWithCurrManager  Attrition  
0             8.0                     7.0                  7.0         No  
1             1.0                     0.0                  0.0        Yes  
2             5.0                     0.0                  3.0         No  
3             7.0                     0.0                  7.0         No  
4             8.0                     3.0                  7.0         No  

[5 rows x 31 columns]

In [35]:
data_available.shape

(2940, 31)

We have 2940 instances and 31 variables including the attrition. 

In [36]:
data_available.isna().sum()/data_available.shape[0] 

hrs                        0.0
absences                   0.0
JobInvolvement             0.0
PerformanceRating          0.0
EnvironmentSatisfaction    0.0
JobSatisfaction            0.0
WorkLifeBalance            0.0
Age                        0.0
BusinessTravel             0.0
Department                 0.0
DistanceFromHome           0.0
Education                  0.0
EducationField             0.0
EmployeeCount              0.0
EmployeeID                 0.0
Gender                     0.0
JobLevel                   0.0
JobRole                    0.0
MaritalStatus              0.0
MonthlyIncome              0.0
NumCompaniesWorked         0.0
Over18                     0.0
PercentSalaryHike          0.0
StandardHours              0.0
StockOptionLevel           0.0
TotalWorkingYears          0.0
TrainingTimesLastYear      0.0
YearsAtCompany             0.0
YearsSinceLastPromotion    0.0
YearsWithCurrManager       0.0
Attrition                  0.0
dtype: float64

There's no Nan values in any of the columns in the available data

In [37]:
data_available.dtypes

hrs                        float64
absences                   float64
JobInvolvement             float64
PerformanceRating          float64
EnvironmentSatisfaction    float64
JobSatisfaction            float64
WorkLifeBalance            float64
Age                        float64
BusinessTravel              object
Department                  object
DistanceFromHome           float64
Education                  float64
EducationField              object
EmployeeCount              float64
EmployeeID                 float64
Gender                      object
JobLevel                   float64
JobRole                     object
MaritalStatus               object
MonthlyIncome              float64
NumCompaniesWorked         float64
Over18                      object
PercentSalaryHike          float64
StandardHours              float64
StockOptionLevel           float64
TotalWorkingYears          float64
TrainingTimesLastYear      float64
YearsAtCompany             float64
YearsSinceLastPromot

In [38]:
data_available.columns

Index(['hrs', 'absences', 'JobInvolvement', 'PerformanceRating',
       'EnvironmentSatisfaction', 'JobSatisfaction', 'WorkLifeBalance', 'Age',
       'BusinessTravel', 'Department', 'DistanceFromHome', 'Education',
       'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender', 'JobLevel',
       'JobRole', 'MaritalStatus', 'MonthlyIncome', 'NumCompaniesWorked',
       'Over18', 'PercentSalaryHike', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
       'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Attrition'],
      dtype='object')

The data is unbalanced with a 83.8% of No and 16.12% of Yes in the response variable

In [39]:
data_available.groupby("Attrition").count()/data_available.shape[0]

hrs  absences  JobInvolvement  PerformanceRating  \
Attrition                                                          
No         0.838776  0.838776        0.838776           0.838776   
Yes        0.161224  0.161224        0.161224           0.161224   

           EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance  \
Attrition                                                              
No                        0.838776         0.838776         0.838776   
Yes                       0.161224         0.161224         0.161224   

                Age  BusinessTravel  Department  ...  NumCompaniesWorked  \
Attrition                                        ...                       
No         0.838776        0.838776    0.838776  ...            0.838776   
Yes        0.161224        0.161224    0.161224  ...            0.161224   

             Over18  PercentSalaryHike  StandardHours  StockOptionLevel  \
Attrition                                                                 
No         0.838776           0.838776       0.838776          0.838776   
Yes        0.161224           0.161224       0.161224          0.161224   

           TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
Attrition                                                             
No                  0.838776               0.838776        0.838776   
Yes                 0.161224               0.161224        0.161224   

           YearsSinceLastPromotion  YearsWithCurrManager  
Attrition                                                 
No                        0.838776              0.838776  
Yes                       0.161224              0.161224  

[2 rows x 30 columns]

In [41]:
data_available.dtypes


hrs                        float64
absences                   float64
JobInvolvement             float64
PerformanceRating          float64
EnvironmentSatisfaction    float64
JobSatisfaction            float64
WorkLifeBalance            float64
Age                        float64
BusinessTravel              object
Department                  object
DistanceFromHome           float64
Education                  float64
EducationField              object
EmployeeCount              float64
EmployeeID                 float64
Gender                      object
JobLevel                   float64
JobRole                     object
MaritalStatus               object
MonthlyIncome              float64
NumCompaniesWorked         float64
Over18                      object
PercentSalaryHike          float64
StandardHours              float64
StockOptionLevel           float64
TotalWorkingYears          float64
TrainingTimesLastYear      float64
YearsAtCompany             float64
YearsSinceLastPromot

In [42]:
data_available.head(10)

hrs  absences  JobInvolvement  PerformanceRating  \
0  7.462297      13.0             3.0                3.0   
1  7.488597       6.0             1.0                3.0   
2  7.116541      16.0             2.0                4.0   
3  9.709125       6.0             3.0                4.0   
4  6.297729       6.0             4.0                3.0   
5  5.742623      13.0             2.0                3.0   
6  5.792075      21.0             2.0                3.0   
7  6.792231      14.0             2.0                3.0   
8  5.601004      17.0             2.0                3.0   
9  6.918927       7.0             2.0                3.0   

   EnvironmentSatisfaction  JobSatisfaction  WorkLifeBalance   Age  \
0                      3.0              3.0              3.0  48.0   
1                      3.0              4.0              4.0  31.0   
2                      1.0              1.0              3.0  36.0   
3                      3.0              3.0              3.0  49.0   
4                      3.0              2.0              3.0  40.0   
5                      4.0              3.0              3.0  52.0   
6                      3.0              4.0              2.0  27.0   
7                      3.0              1.0              2.0  49.0   
8                      4.0              3.0              3.0  47.0   
9                      4.0              1.0              3.0  29.0   

      BusinessTravel              Department  ...  Over18  PercentSalaryHike  \
0      Travel_Rarely  Research & Development  ...       Y               13.0   
1  Travel_Frequently  Research & Development  ...       Y               17.0   
2      Travel_Rarely  Research & Development  ...       Y               24.0   
3      Travel_Rarely  Research & Development  ...       Y               22.0   
4  Travel_Frequently  Research & Development  ...       Y               13.0   
5      Travel_Rarely  Research & Development  ...       Y               11.0   
6      Travel_Rarely                   Sales  ...       Y               13.0   
7      Travel_Rarely  Research & Development  ...       Y               11.0   
8      Travel_Rarely  Research & Development  ...       Y               19.0   
9      Travel_Rarely  Research & Development  ...       Y               12.0   

  StandardHours  StockOptionLevel  TotalWorkingYears TrainingTimesLastYear  \
0           8.0               1.0               18.0                   2.0   
1           8.0               1.0                3.0                   2.0   
2           8.0               0.0                6.0                   6.0   
3           8.0               0.0               25.0                   2.0   
4           8.0               1.0                8.0                   3.0   
5           8.0               1.0                9.0                   1.0   
6           8.0               1.0                6.0                   2.0   
7           8.0               1.0               16.0                   3.0   
8           8.0               3.0               27.0                   3.0   
9           8.0               0.0               11.0                   4.0   

   YearsAtCompany YearsSinceLastPromotion YearsWithCurrManager  Attrition  
0             8.0                     7.0                  7.0         No  
1             1.0                     0.0                  0.0        Yes  
2             5.0                     0.0                  3.0         No  
3             7.0                     0.0                  7.0         No  
4             8.0                     3.0                  7.0         No  
5             5.0                     1.0                  4.0         No  
6             6.0                     1.0                  4.0         No  
7            15.0                     5.0                 11.0         No  
8             5.0                     1.0                  0.0         No  
9             3.0                     1.0                  2.0         No  


In [44]:
categorical_variables = ['JobInvolvement', 'PerformanceRating','EnvironmentSatisfaction', 'JobSatisfaction',
                          'WorkLifeBalance','BusinessTravel', 'Department', 'Education','EducationField',
                          'EmployeeID','Gender', 'JobLevel','JobRole','MaritalStatus','StockOptionLevel','Attrition']
numeric_variables = ['hrs','absences','Age','DistanceFromHome','MonthlyIncome','NumCompaniesWorked',
                     'PercentSalaryHike', 'StandardHours','TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
       'YearsSinceLastPromotion', 'YearsWithCurrManager']

constant_variables = ["EmployeeCount",'Over18','StandarHours']

In [45]:
data_available[categorical_variables].nunique()

JobInvolvement                4
PerformanceRating             2
EnvironmentSatisfaction       4
JobSatisfaction               4
WorkLifeBalance               4
BusinessTravel                3
Department                    3
Education                     5
EducationField                6
EmployeeID                 2940
Gender                        2
JobLevel                      5
JobRole                       9
MaritalStatus                 3
StockOptionLevel              4
Attrition                     2
dtype: int64